Add package API (async transfers)

sevein opened the pull request:
[RDSSARK-474](https://jiscdev.atlassian.net/browse/RDSSARK-474)

New package API - it allows to create new packages.

It's an alternative to the [existing methods](https://gist.github.com/sevein/5615e0df0ff9645f17ed06672a271633) to start a transfer. JiscRDSS's channel adapter (deposit packages directly into `activeTransfers/*`) and automation tools (uses existing `/api/{start_transfer,approve,unapproved}` APIs) could adopt this new API but the existing functionality isn't going to stop working. This PR updates the transfer browser widget to adopt the new API - before it used internal endpoints (`/filesystem/transfer/`, `/mcp/list`, `/mcp/execute`).

Request example:

```
$ curl \
	-v \
	-d "@send-transfer.json" \
	-H "Content-type: application/json" \
	-H "Authorization: ApiKey test:test" \
	-X POST \
		http://127.0.0.1:62080/api/v2beta/package/
```

This is the document we sent in the payload:

```json5
{
  "name": "Test12345",
  // The value of `path` is the result of `base64(location-uuid:relative-path)`
  "path": "MWQ0ZmU3NTctNGM2ZS00MmQxLWE5Y2ItZmQ4YTBkMjNlZWRkOi9ob21lL2FyY2hpdmVtYXRpY2EvYXJjaGl2ZW1hdGljYS1zYW1wbGVkYXRhL1NhbXBsZVRyYW5zZmVycy9JbWFnZXMvcGljdHVyZXM="
}
```

It starts the transfer without approval - as MCPServer knows when the files have been copied and the package is ready for processing. `POST /package` returns the UUID of the package/transfer immediately.

sevein committed:
MCPServer: delete dead code

sevein committed:
MCPServer: add SS timeout config

We're using the SS client in upcoming commits. This commit provides the
configuration attribute that is necessary by the client.

sevein committed:
Dashboard: use package endpoint in transfer browser

The new `POST /api/package` endpoint targets a single package. This commit
updates the transfer service in the transfer browser application to use the
new endpoint and generate one request per directory chosen by the user.

sevein committed:
Dashboard: delete /filesystem/transfer

With the transfer browser widget now using the new `/package` API this code
becomes unused and it can be removed.

sevein committed:
MCPServer: add packageCreate RPC method

sevein committed:
Dashboard: new /package API

sevein commented on src/MCPServer/lib/archivematicaMCP.py:
```diff
@@ -659,6 +660,13 @@ def _except_hook_log_everything(exc_type, exc_value, exc_traceback):
     t.daemon = True
     t.start()
+
+    # TODO: how many workers should we have? Should this be configurable?
```
I'd like to get your thoughts here!

helenst replied:
I guess this comes down to how many package transfers might need to be started simultaneously? (at least for now with that being the only endpoint served by the workers). I don't know what is a good number to choose there but making it configurable seems sensible to me.

helenst commented on src/MCPServer/lib/package.py:
```diff
@@ -0,0 +1,391 @@
+                       <uuid of location>:<full path in location>
+    :param str relative_destination: Path relative to the currently processing
+                                     space to move the files to.
+    :returns: Tuple of (boolean error, message)
```
It doesn't return anything. Given that the calling function doesn't expect a return value, I am guessing it is the docstring that is incorrect?

helenst commented on src/MCPServer/lib/package.py:
```diff
@@ -0,0 +1,391 @@
+        try:
+            Transfer.objects.get(pk=args[0])
+
+            #
```
does this also need adding to the TODO list in the main PR comments?

sevein replied:
I had some code here that was updating the `Transfer.state` FSM. We decided to think about it a bit further so I deleted it. I'm going to add a comment to make that clear. I wanted to leave the decorator in place because I thought it would be clearer to have it even if it's a no-op right now.

sevein committed:
Dashboard: delete shebang from tests

helenst reviewed (commented):
I guess there are still things to be done here but so far this seems to be working and looks good to me - just a couple of minor comments.

sevein commented:
@helenst I've addressed your comments, thanks!

jrwdunham commented on src/MCPServer/lib/RPCServer.py:
```diff
@@ -31,60 +25,62 @@
 from linkTaskManagerChoice import choicesAvailableForUnits
```
Nitpick: shouldn't this go below the third-party imports?

jrwdunham commented on src/MCPServer/lib/worker_util.py:
```diff
@@ -0,0 +1,55 @@
+
+
+def unpickle_payload(fn):
+    """Worker handle decorator to unpickle the incoming payload.
```
Nitpick: should "Worker handle" here and "Working handle" below both be "Worker handler" as above?

jrwdunham commented on src/MCPServer/lib/RPCServer.py:
```diff
@@ -31,60 +25,62 @@
+@pickle_result
+@unpickle_payload
+@log_exceptions(LOGGER)
+def job_approve_handler(payload, gearman_worker, gearman_job):
```
This is very cool. Although, it's a bit confusing because these Gearman task handlers don't have the expected signature. Also, these 3 funcs take args that are never used. Could you replace all of the `gearman_worker, gearman_job` in these 3 funcs with just `*args`? Would that make it more readable? Same question applies to the handler in lib/worker.py.

jrwdunham commented on src/MCPServer/lib/package.py:
```diff
@@ -0,0 +1,391 @@
+
+        except Exception:
+            pass
+        raise err
```
It's possible to raise `None` here, which we should not allow I think. Pylint says "Raising NoneType while only classes or instances are allowed (raising-bad-type)"

jrwdunham commented on src/MCPServer/lib/package.py:
```diff
@@ -0,0 +1,391 @@
+}
+
+
+def _file_is_an_archive(file):
```
Nitpick: shouldn't use `file` as a var name.

jrwdunham commented on src/MCPServer/lib/package.py:
```diff
@@ -0,0 +1,391 @@
+        error = 'Filepath ' + filepath + ' does not exist.'
+    # check if is file or directory
+    # check for trickery
+    try:
```
This is odd. Why not:

    def _check_filepath_exists(filepath):
        if filepath == '':
            return 'No filepath provided.'
        if not os.path.exists(filepath):
            return 'Filepath ' + filepath + ' does not exist.'
        if '..' in filepath:  # check for trickery
            return 'Illegal path.'
        return None

jrwdunham commented on src/MCPServer/lib/package.py:
```diff
@@ -0,0 +1,391 @@
+                                     space to move the files to.
+    """
+    processing_location = storage_service.get_location(purpose='CP')[0]
+    transfer_sources = storage_service.get_location(purpose='TS')
```
Too bad we need to perform 2 requests here. Another argument for revising the SS API.

jrwdunham replied:
Also too bad that this is basically copy-pasted from filesystem_ajax/views.py. It would be nice if we could consolidate these and then just change how the func is called in filesystem_ajax/views.py and catch the exceptions raised here.

jrwdunham commented on src/MCPServer/lib/package.py:
```diff
@@ -0,0 +1,391 @@
+def _submit_create_package_internal_bg_job(id_, name, type_, path):
+    """Submit package creation background job to queue.
+
+    It returns the job handle assigned by Gearman.
```
Why does it return anything if the return value is never used?

jrwdunham commented on src/MCPServer/lib/package.py:
```diff
@@ -0,0 +1,391 @@
+    return request.job.handle
+
+
+def _create_temporary_directory():
```
Again, I know this is taken from elsewhere in AM, but this seems like a good candidate for a context manager which yields the tmpdir and calls `shutil.rmtree` on it at the end so the caller doesn't have to. Maybe an enhancement that is out of scope for this work. Just an idea.

jrwdunham commented on src/MCPServer/lib/package.py:
```diff
@@ -0,0 +1,391 @@
+    if _file_is_an_archive(path):
+        transfer_dir = tmpdir
+        p = path.split(':', 1)[1]
+        logger.debug('Package %s: found a zip file, splitting path %s', id_, p)
```
Seems odd to call it a "zip" file when it could be different types of archive. Also, doesn't it sound odd to say we're splitting the result of the split? Shouldn't `p` here be `path`?

jrwdunham commented on src/MCPServer/lib/package.py:
```diff
@@ -0,0 +1,391 @@
+    else:
+        path = os.path.join(path, '.')  # Copy contents of dir but not dir
+        transfer_dir = os.path.join(tmpdir, name)
+        filepath = os.path.join(tmpdir, name)
```
Again, I see lots of this is copy/paste from filesystem_ajax/views.py, but couldn't this just be `transfer_dir = filepath = os.path.join(tmpdir, name)`?

jrwdunham commented on src/MCPServer/lib/package.py:
```diff
@@ -0,0 +1,391 @@
+    src = dest
+    dest = dest.replace(django_settings.SHARED_DIRECTORY, '%sharedPath%', 1)
+    if dest.endswith("/"):
+        dest = os.path.join(dest, os.path.basename(src))
```
This seems really weird. I'd like to understand if this code path ever gets run and what is expected to happen when it does, and why. From the snippet below, it looks like it would do nothing...

jrwdunham replied:
>>> dest = '/a/b/c/d/e/f/'
    >>> src = dest
    >>> dest = dest.replace('/a/b', '%sharedPath%', 1)
    >>> dest
    '%sharedPath%/c/d/e/f/'
    >>> src
    '/a/b/c/d/e/f/'
    >>> if dest.endswith('/'):
    ...     dest = os.path.join(dest, os.path.basename(src))
    ... 
    >>> dest
    '%sharedPath%/c/d/e/f/'

jrwdunham commented on src/MCPServer/lib/settings/common.py:
```diff
@@ -168,3 +172,5 @@
 LIMIT_GEARMAN_CONNS = config.get('limit_gearman_conns')
 RESERVED_AS_TASK_PROCESSING_THREADS = config.get('reserved_as_task_processing_threads')
 SEARCH_ENABLED = config.get('search_enabled')
+STORAGE_SERVICE_CLIENT_TIMEOUT = config.get('storage_service_client_timeout')
```
Is this setting ever used in the MCPServer? I searched for it and couldn't find it. Is it needed?

sevein replied:
Yes, it's an implicit requirement when you import storageService because that module looks up `django_settings`. I realize it's not great, I will think about it.

jrwdunham commented on src/MCPServer/lib/worker_error.py:
```diff
@@ -0,0 +1,26 @@
+
+class RPCServerError(Exception):
+    """Base exception of RPCServer."""
+    pass
```
Nitpick: you don't need `pass` if you have a docstring.

jrwdunham commented on src/dashboard/src/contrib/mcp/client.py:
```diff
@@ -56,3 +63,35 @@ def notifications(self):
+        if response.state == gearman.JOB_COMPLETE:
+            return cPickle.loads(response.result)  # Transfer ID (pickled)
+        elif response.state == gearman.JOB_FAILED:
+            raise RPCError('MCPServer returned an error (check the logs)')
```
Are other `state` values not also possible here? What about

    JOB_UNKNOWN  = 'UNKNOWN'  # Request state is currently unknown, either unsubmitted or connection failed
    JOB_PENDING  = 'PENDING'  # Request has been submitted, pending handle
    JOB_CREATED  = 'CREATED'  # Request has been accepted

sevein replied:
I think they are but I don't know what I would do with them. We have other RPC methods doing the same and it's never been a problem. We're hoping to use gRPC soon but the first iteratin is just reusing this ugly RPC mechanism we invented on top of Gearman. It's ugly but I wouldn't try to make it much better, kind of reinventing the wheel when we know gRPC is awesome for this purpose. I guess this is one of these situations where I can use the "walking skeleton" card hoping that we can ignore this for now... ;-)

jrwdunham reviewed (commented):
I have some comments. I've confirmed that I can use the example curl command to start different types of transfers. I guess that the path to the packages resources will be changed, but for now the description of this PR is incorrect in that you have to make a POST request to `/api/v2beta/package/` and not the stated `/api/package/`.

I also noticed that if you start a Bag or Zipped Bag transfer, the first half dozen or so micro-services are listed under the group "Approve transfer" which doesn't really make sense any more because approval is gone. I think this should probably be changed and it will require a migration I think.

sevein committed:
Dashboard: update /api/transfer/start_transfer

In a previous commit we have added to MCPServer the ability to create packages
and start tranfers. This commit updates `/api/transfer/start_transfer/` and the
underlying views to make use of this new functionality. In particular, it uses
the `auto_approve=False` mode to maintain backward compatibility - this method
also blocks until the files are copied/moved to the activeTransfer directory.

sevein commented:
For now only merged in JiscRDSS. The version here is older and needs to be updated.
